### DISCUSSION TITLE
Create schema with sharded session using sql function

### DISCUSSION THREAD

**Discussion by fabiend-s** [sqlalchemy/sqlalchemy#8611] [Usage Questions] [answered]
Hello,

In my current project, we are using SQLAlchemy 1.4.29 + psycopg 2.8.6 with a PostgreSQL 9.2.24 on Centos 7.

Our database is defined with multiple schemas like this :
* 1 'public' schema containing common Tables with data for the whole application and the whole database
  * ex: money, house_type, country ...
  * country will list every known country and have a "country.schema_" column to know the country's schema name
* X schemas (one for each country) containing the same Tables in each, but with data specific to the given country, ex:
  * schema "usa" with usa.people, usa.house, usa.mailing_list_view
  * schema "france" with france.people, france.house, france.mailing_list_view

It often happens that a Table of those schema points to a Table in the 'public' schema
* ex: <usa, france, ...>.house.id_type = ForeignKey(house_type.id)

A country schema can be created by the user in the application. This includes :
1. the creation of the schema in the database
2. the call of SQL functions to :
2.a. create Tables in it (people, house ...) ("`select create_tables_country()`")
2.b. create Views in it (mailing_list_view ...) ("`select create_views_country()`")

We also need to be able to export SOME of the data from different schemas to a new database. So we need to :
1. create a new database
2. create the 'public' schema with the minimum data in it
3. create only the specified schemas (ex: usa)
4. export the specified data from, ex: usa.house.id IN (1, 3 ...) ...
5. so we have to export the some data from the 'public' because of the ForeignKeys, ex: if usa.house.id_type = 3, we need to export (public.)house_type.id = 3, but NOT (public.)house_type.id = 2

All of this should be atomic, so we want to use only one session to be able to commit / rollback the whole operation at once.
To do this, we used to call session.execute("set search_path TO <usa/france/...>, public").

We ran into the problem that some of the exported data were not well exported :
if we export usa.house.id = 2 AND france.house.id = 2, the session.identity_map can't differentiate them, so it CAN happen that we export a house from the original database schema "france" to the exported database schema "usa"

To resolve this, we switched to using ShardedSession and ShardedQuery using this example : https://docs.sqlalchemy.org/en/14/_modules/examples/sharding/separate_tables.html
To do so we:
1. create a basic session and request the public.country Table to get all the existing schemas of the database (public is always created) and close this session
2. create an engine (shard) for each schema with `connect_args={'options': '-csearch_path={}'.format(schema_name)}` and save them in a dict 'shards' {schema_name: engine}
3. create a ShardedSession using sessionmaker() + sessionmaker.configure(), with :
3.a. `info={'schema': desired_schema, shards_dict='shards'}`
3.b. id_chooser and execute_chooser using the info to return correct shard

This indeed resolves the problem of differentiating the same key accross multiple schemas, thanks to the identity_token added in the object's identity_key.

BUT we are now unable to create a new schema. Before switching to ShardedSession we used to do (and it worked in less than a minute) :
```
session.execute(sqlalchemy.schema.CreateSchema("france"))  # 1.
session.execute("set search_path to 'france,public'")  # 2.
session.execute(sqlalchemy.func.create_tables_country())  # 3.
session.execute(sqlalchemy.func.create_views_country())  # 4.
```

Now there are some "minor" changes :
1. Create the schema
1.a. call session.execute(sqlalchemy.schema.CreateSchema("france"))
1.b. create a new engine (shard) and set its search_patch (like in creating a new session 2.)
1.c. add this engine to our 'shards' dictionnary AND to the session's known engine using session.bind_shard("france", new_shard)
2. update the session's info to set its 'schema' to 'france,public' (the id_chooser and execute_chooser will return the newly created engine this way).

This seems to works well up to this stage, but the next step (3. call session.execute(sqlalchemy.func.create_tables_country())) just ... stop working mid-way ?
postgreSQL does not seem to log anything, and we have to just kill the process as we waited more than 5 minutes.

Do you know if creating a new schema is possible with ShardedSession ? And if we can just create a new engine/shard and use it this way ?

Regards,
Fabien D

**Comment by zzzeek** [comment, marked as answer]
the ShardedSession is going to want to route specific operations to specific "shards".   so when running an operation that is "cross-shard", you need to isolate the database connection from the Session itself.   , so that the event hook used by ShardedSession is not invoked.

I would recommend doing it as:

```py
     session.connection().execute(sqlalchemy.schema.CreateSchema("france"))
```

Another way that might be more straightforward to control is to not use sharded session and assign identity token directly

```py

from sqlalchemy import inspect
obj = MyObject(...)

inspect(obj).identity_token = "france"

regular_session.add(obj)
```

that should persist it with that identity token (haven't tried but this looks like what shardedsession does).

**Reply by fabiend-s** [reply]
Thanks for the quick and useful reply.

We tried the first proposition (session.connection().execute(...) and it works nicely to create the schema and use the sql function.
Sadly, the same problem arises now at the first query on this schema.

The problem is that the sharded engines seem to not know of the newly create schema :
```
from sqlalchemy import inspect
inspect(session.get_bind(shard_id='public/france')).get_schema_names()
```
returns only [u'information_schema', u'public'] ;
while `inspect(session.connection()).get_schema_names()`
returns [u'information_schema', u'public', u'france']

as a result we still can't use the newly created shard / schema.

Do you know if there is any way to update schemas in engines / connection ?

**Reply by zzzeek** [reply]
OK well you said you wanted everything inside of one transaction.  if you are putting your "CREATE SCHEMA" in that non-committed transaction, then it wont be visible from other connections.   you would need to commit the transaction as session.get_bind() returns an Engine that will give you other connections.

**Reply by fabiend-s** [reply]
Ok, that's what I was afraid of. I won't be able to run some tests quickly. I tried quickly to create another session just to create this new schema (commit and close the session) but it did not work (I didn't think to check if the sharded session knew about the new schema).

I won't be able to run tests quickly, but I'll try some things with this new insights.

The last way I thought of was to create all schemas (or only the desired ones if I can have this information soon enough) with a regular session (commit, close it), then call the ShardedSession which would know about all of them, and finally delete the undesired schemas (within the ShardedSession if possible).

I'll let you know what I was able to do to get things done.

**Reply by zzzeek** [reply]
> Ok, that's what I was afraid of. I won't be able to run some tests quickly. I tried quickly to create another session just to create this new schema (commit and close the session) but it did not work (I didn't think to check if the sharded session knew about the new schema).

you mean you created the schema in a separate transaction, and your current, ongoing transaction could not see the schema?   that doesnt sound right and I can't reproduce that.  assume "conn1" and "conn2" are two separate sessions:

```sql
conn1 test=> BEGIN;
BEGIN
conn1 test=*> create table foo (id integer);
CREATE TABLE

conn2 test=> begin;
BEGIN
conn2 test=*> create schema my_schema;
CREATE SCHEMA
conn2 test=*> commit;
COMMIT

-- conn1 sees new schema
conn1 test=*> create table my_schema.bar(id integer);
CREATE TABLE
conn1 test=*> commit;
COMMIT


```

> 
> The last way I thought of was to create all schemas (or only the desired ones if I can have this information soon enough) with a regular session (commit, close it), then call the ShardedSession which would know about all of them, and finally delete the undesired schemas (within the ShardedSession if possible).

That's how I would do this myself.   mixing DDL with data operations is a recipe for confusion.

**Reply by fabiend-s** [reply]
For sake of completion, I just ran again the same test to create the schema and this time it seems to work :
```
def create_session()
  shards = create_engine(...)
  session_maker = sessionmaker(ShardedSession)
  session_maker.configure(shards ...)
  return session_maker()

session_1 = create_session()
session_2 = create_session()
session_2.execute(CreateSchema("france"))
session_2.commit()
session_2.close()

inspect(session_1.connection()).get_schema_names()  # this time the result includes "france" ..?
```

Anyway we now try to create the required schemas before, and only then create the ShardedSession (with all the required schemas) to populate it.
The new problem is that when we populate the "usa" schema using the "usa" shard, we also populate the "public" schema, for example to add the public.house_type = 2 if required by the usa.house.id_type.
```

But the "france" shard does not know about it !
# Define the session to work on "usa" schema/shard with session.info and the X_chooser() functions
session.query(HouseType).filter_by(id=2)  # returns an object

# Define the session to work on "france" schema/shard
session.query(HouseType).filter_by(id=2)  # returns nothing

# Define again the session to work on "usa" schema/shard
session.query(HouseType).filter_by(id=2)  # returns again an object
```

I thought the whole point of the horizontal sharding was to be able to keep track of datas from mutiple schemas via the use a central session. I understand the different shards are using different engines so what I excepted can be just not possible.

The good news is that we finally found a way to what we wanted. We go back to the classical session and just call :
```
session.execute("SET SEARCH_PATH TO %s", schema)  # What we did before
session.expunge_all()
```

This lead to the session cleaning its identity_map, so when we change the schema from "usa" to "france" and request house.id = 2 in both, the session don't think it is the house it already knows.

Don't know how to flag this question as we seem to have found a solution (need a bit more testing to be sure there is no other side effects but it seems to be good), but it is not with the sharded session at all.

**Reply by zzzeek** [reply]
horizontal sharding knows about engines, not specifically schemas.   so you would want the different engines to be configured for their specific schema and to eliminate any ability for them to cross-talk with each other, which means setting search path on those engines to not include "public".

**Reply by fabiend-s** [reply]
> [...]  so you would want the different engines to be configured for their specific schema [...] .

That's what I did, but indeed still letting all shards know about the "public" schema :
`connect_args={'options': '-csearch_path={},public'.format(schema_name)}`

> [...] which means setting search path on those engines to not include "public".

It was another way I thought to try indeed.
But that would have mean to be able to differentiate the shards according to Table mentioned in the query, and I was afraid of side-effects in some requests which contains both "public" and "country" schema like :
`SELECT house.address, house_type.name FROM house JOIN house_type ON house.id_type = house_type.id WHERE house.id = 2`

I could have add explicitly the {'schema' : 'public'} information in the public classes' model for instance, but then again 'public' in a request would not mean it is obviously in the 'public' schema that I want to work.

This is where we stopped and took a step back, read a bit more on the session.identity_map and just thought "well, what if we just clear this kind of cache", tried it and it worked without any other change to do in the code (so no more ShardedSession etc.).

Maybe we missed a step somewhere and the ShardedSession could still be a better solution, but right now it seems much more complex and add nothing to our need, while just calling "set search_path to <schema>" finally does the trick with PostgreSQL, given that you "reset" the object "cached in" the session when doing so.

**Reply by zzzeek** [reply]
it looks like you thought it through so if you have somehting that works, the simplest approach is best.

**Reply by fabiend-s** [reply]
Yes, still thanks you for your quick and precise answers as it allows me to really understand and try things.

I'll tag your original post as "Mark as answer" to close this thread as I don't think I'll go more into this subject.
